Rules/UseCompatibleTypes.cs

```diff
@@ -0,0 +1,854 @@
+        /// </summary>
+        private void ProcessDirectory(string path, List<string> compatibilityList)
+        {
+            var jsonFiles = Directory.EnumerateFiles(path, "*.json");
+            if (jsonFiles == null)
+            {
+                return;
+            }
+
+            foreach (string file in jsonFiles)
+            {
+                string fileNameWithoutExt = Path.GetFileNameWithoutExtension(file);
+                if (!compatibilityList.Contains(fileNameWithoutExt, StringComparer.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                dynamic deserialized = JObject.Parse(File.ReadAllText(file));
+                psTypeMap[fileNameWithoutExt] = GetTypesFromData(deserialized);
+            }
+        }
+
+        /// <summary>
+        /// Get a hashset of full type names from a deserialized json file.
+        /// </summary>
+        private HashSet<string> GetTypesFromData(dynamic deserializedObject)
+        {
+            HashSet<string> types = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            dynamic typeList = deserializedObject.Types;
+            foreach (dynamic type in typeList)
+            {
+                string name = type.Name.ToString();
+                string nameSpace = type.Namespace.ToString();
+                string fullName = nameSpace + "." + name;
+                types.Add(fullName);
+            }
+            return types;
+        }
+
+        /// <summary>
+        /// Create the type accelerator map from json file in Settings directory.
+        /// </summary>
+        private void CreateTypeAcceleratorMap(string path)
+        {
+            string typeAccFile = Path.Combine(path, typeAccFileName);
+            dynamic deserialized = JObject.Parse(File.ReadAllText(typeAccFile));
+
+            foreach (dynamic typeAcc in deserialized)
+            {
+                typeAcceleratorMap.Add(typeAcc.Name.ToString(), typeAcc.Value.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Set up type map from the latest desktop version of PowerShell.
+        /// </summary>
+        private void CreateReferenceTypeMap(string path, string fileName)
+        {
+            string[] typeFile = Directory.GetFiles(path, fileName);
+            dynamic deserialized = JObject.Parse(File.ReadAllText(typeFile[0]));
+            referenceMap = GetTypesFromData(deserialized);
+        }
+
+        ///<summary>
+        /// Separate type from New-Ojbect command, pass to GetFullNames to retrieve full type name.
+        ///</summary>
+        private List<fullTypeNameObject> RetrieveFullTypeName(dynamic typeObject)
+        {
+            List<fullTypeNameObject> fullNameObjectList = new List<fullTypeNameObject>();
+
+            // Check to see if our object is a CommandAst. 
+            if (typeObject is CommandAst)
```

Comment:
A large nested `if` statement like this should be refactorable with `return`s hopefully